File: array/maximum_subarray.go
package main

import "fmt"

func calc(arr []int) int {
	max := 0
	out := 0
	for i := 0; i < len(arr); i++ {
		if arr[i] < max+arr[i] {
			max += arr[i]
		} else {
			max = arr[i]
		}
		if out < max {
			out = max
		}
	}
	return out
}

func main() {
	fmt.Println(calc([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}))
}

File: array/maximum_subarray.cc
#include <vector>
#include <algorithm>
#include <stdio.h>

//std::vector calc(std::vector<int> nums) {
int calc(std::vector<int> nums) {
    int N = nums.size();

    int dp = 0; // will reset if dp+nums[i] is smaller than nums[i]
    int ans = 0;
    int curMax = 0;
    for (int i = 0; i < N; i++) {
        curMax = std::max(nums[i], curMax+nums[i]);
        ans = std::max(ans, curMax);
        printf("%d : %d %d\n", i, curMax, ans);
    }

    return ans;
}

int main() {
    std::vector<int> nums{-2,1,-3,4,-1,2,1,-5,4};
    auto out = calc(nums);
    printf("%d\n", out);
}
